### PR TITLE
pkg-config:  Define SEQAN_HAS_* flags in pkg-config file

### PR DESCRIPTION
**Pull request by kdmurray91**
Hi all,

I might be wrong, but would it be better if the `pkg-config` file for seqan defined `SEQAN_HAS_GZIP=1` etc? I would think that the pkg-config files should define these for the system, instead of developer code having to manually define these.

(`git clean -fdx` removed a vim temp file, which I assume was not really required)

Cheers,
K

### CODE REVIEW

**Comment by h-2** [comment]
@kdmurray91 thanks for the PR!

it is true that the pkgconfig is rather minimal now, but I am not sure this is the correct way to go about it. Because now it depends on the system building the tarball, whether e.g. zlib is set as a dependency or not. But this should be variable, depending on the system where the pkgconfig is used.

Can we somehow make this dynamic?

If not, then why don't we just hardcode it, independent of the operating system that packages it? At least for zlib and bzip2.

For the c++-standard I am reluctant to force it on people, because it might conflict with someone setting a different, but compatible flag (e.g. `-std=c++1z` or `-std=gnu++14`). For CMake we check this...

**Comment by kdmurray91** [comment]
Hi @h-2,

I tried to mirror the logic of FindSeqan.cmake as much as possible when creating the pkgconfig file. Do you have stats on what proportion of users **don't** have zlib & bzip2 enabled? I would imagine it being pretty small, in which case hard-coding these makes sense.

 In any case, one can override these flags with consecutive definitions (```g++ -o tst `pkg-config --cflags seqan2` -UFLAG -DFLAG=0```  would override FLAG if given in pkgconfig). And `pkgconfig --libs` ignored completely given seqan is header only.

Regarding c++ standard, that's a good point. But, one can override the `-std` set in seqan's pkgconfig by setting a different standard later in the compilation command line (e.g. `CXXFLAGS = $(shell pkg-config --cflags seqan2) -std=gnu++14` in a makefile would cause the gnu++14 standard to be used AFAICT. I'm happy to remove this from pkg-config though.

**Comment by h-2** [comment]
>  Do you have stats on what proportion of users don't have zlib & bzip2 enabled? I would imagine it being pretty small, in which case hard-coding these makes sense.

Well the thing is that they also need the headers installed which are not default on some systems (including debian I think). But I am still fine with requiring this of all users and hard-coding it. Then we don't need any changes to the cmake-files. But, I am still unsure: will pkg-config actually check for the availability of the headers, or does it just assume they are there?

> Regarding c++ standard, that's a good point. But, one can override the -std set in seqan's pkgconfig by setting a different standard later in the compilation command line

If it is the default for pkg-config based projects to add their flags after pkg-config's ones that would be totally ok. To be honest no one here uses pkgconfig so we will just trust your judgement on what's best :smile:

**Comment by kdmurray91** [comment]
> Well the thing is that they also need the headers installed which are not default on some systems (including > debian I think). But I am still fine with requiring this of all users and hard-coding it.  Then we don't need any > changes to the cmake-files.

OK, that sounds reasonable. Will modify the PR accordingly.

> But, I am still unsure: will pkg-config actually check for the availability of the headers, or does it just assume > they are there?

AFAIK, pkg-config doesn't check for headers. If the dependencies have pkg-config files (like zlib does), then you can specify dependencies that are checked for. But you can't if pkg-config files don't exist (like bz2).

> If it is the default for pkg-config based projects to add their flags after pkg-config's ones that would be totally > ok. To be honest no one here uses pkgconfig so we will just trust your judgement on what's best :smile:

I only know what I do, which is to override CLI flags if required. I think either action is OK. If there are any hard-core autotools/pkgconfig users you know of, it might be worth pinging them. We can always leave it out to be sure. That would prevent user code breaking, as any code that has not manually added -std=c++14 would not compile currently anyway. I might just do that.

**Comment by h-2** [comment]
@kdmurray91 ok, feel free to update this as you see fit!

**Comment by kdmurray91** [comment]
@h-2 is there a deadline for 2.3? I'm quite busy right now, but will get this done before the 2.3 deadline.

**Comment by h-2** [comment]
@kdmurray91 well, we are long overdue, but the CI just broke so it will definitely not happen this week. I do really want to release end of next week though. Thank you for your contributions :+1:

**Comment by kdmurray91** [comment]
@h-2 Sorry for the delay, but I think I've addressed you comments now.

File: util/cmake/SeqAnBuildSystem.cmake [hunk 1]
```diff
@@ -367,6 +367,9 @@ macro (seqan_setup_library)
                  DESTINATION share/doc/seqan)
         # Install pkg-config file, except on Windows.
         if (NOT CMAKE_SYSTEM_NAME MATCHES Windows)
+            # Find packages and set flags appropriately
+            set (SEQAN_PKGCFG_DEPS "${SEQAN_PKGCFG_DEPS} zlib")
+            set (SEQAN_PKGCFG_LIBS "${SEQAN_PKGCFG_LIBS} -lbz2")
```
Comment (h-2) [inline util/cmake/SeqAnBuildSystem.cmake]: can we also hardcode these?

File: util/pkgconfig/seqan.pc.in [hunk 2]
```diff
@@ -6,5 +6,6 @@ Name: @CMAKE_PROJECT_NAME@
 Description: C++ library for biological sequence analysis
 URL: http://www.seqan.de
 Version: @SEQAN_VERSION_STRING@
-Requires: zlib
-Cflags: -I${includedir}
+Requires: @SEQAN_PKGCFG_DEPS@
+Cflags: -I${includedir}  @SEQAN_PKGCFG_CFLAGS@ -DSEQAN_HAS_ZLIB=1 -DSEQAN_HAS_BZIP2=1 -DSEQAN_HAS_OPENMP=1
```
Comment (h-2) [inline util/pkgconfig/seqan.pc.in]: We don't need ` -DSEQAN_HAS_OPENMP=1`, anymore, the code only relies on `_OPENMP`.

But to be able to use this we would instead need `-fopenmp`. Do you think it makes sense to hardcode this? It means it won't work on clang < 3.8 .

File: util/cmake/SeqAnBuildSystem.cmake (hunk 1 above)
Comment (kdmurray91) [inline util/cmake/SeqAnBuildSystem.cmake]: Sure. These are hard-coded in essense, but I can just move their definition to the pkg-config file.

File: util/pkgconfig/seqan.pc.in (hunk 2 above)
Comment (h-2) [inline util/pkgconfig/seqan.pc.in]: @kdmurray91 Ok, thanks! Your patches are much appreciated!

File: util/cmake/SeqAnBuildSystem.cmake (hunk 1 above)
Comment (kdmurray91) [inline util/cmake/SeqAnBuildSystem.cmake]: Done

File: util/pkgconfig/seqan.pc.in (hunk 2 above)
Comment (kdmurray91) [inline util/pkgconfig/seqan.pc.in]: Cool, I think this is ready to roll now. Would be worth double checking by hand that it all work at your end (on your build boxes), as I've only checked this on a couple of systems, with recent pkg-config and cmake versions.

**Pull request merged**